skimage/future/_thin_plate_splines.py

```diff
@@ -0,0 +1,323 @@
+    ...                         0, 1, 2, 3, 4,
+    ...                         0, 1, 2, 3, 4,
+    ...                         0, 1, 2, 3, 4,
+    ...                         0, 1, 2, 3, 4])
+    >>> np.allclose(yy_trans, expected_yy)
+    True
+    """
+
+    def __init__(self):
+        self._estimated = False
+        self.spline_mappings = None
+        self.src = None
+
+    def __call__(self, coords):
+        """Estimate the transformation from a set of corresponding points.
+
+        Parameters
+        ----------
+        coords : (N, 2) array_like
+            x, y coordinates to transform
+
+        Returns
+        -------
+        transformed_coords: (N, D) array
+            Destination coordinates
+        """
+        if self.spline_mappings is None:
+            raise ValueError(f"{self.spline_mappings}. Compute the `estimate`")
+        coeffs = self.spline_mappings
+        coords = np.array(coords)
+
+        if not coords.ndim == 2 or coords.shape[1] != 2:
+            raise ValueError("Input `coords` must have shape (N, 2)")
+
+        radial_dist = self._radial_distance(coords[:, 0], coords[:, 1])
+
+        x_warp = self._spline_function(
+            coords[:, 0], coords[:, 1], radial_dist, coeffs[:, 0]
+        )
+        y_warp = self._spline_function(
+            coords[:, 0], coords[:, 1], radial_dist, coeffs[:, 1]
+        )
+        return np.vstack([x_warp, y_warp]).T
+
+    @property
+    def inverse(self):
+        raise NotImplementedError("This is yet to be implemented.")
+
+    def estimate(self, src, dst):
+        """Estimate optimal spline mappings between source and destination points.
+
+        Parameters
+        ----------
+        src : (N, 2) array_like
+            Control points at source coordinates.
+        dst : (N, 2) array_like
+            Control points at destination coordinates.
+
+        Returns
+        -------
+        success: bool
+            True indicates that the estimation was successful.
+
+        Notes
+        -----
+        The number N of source and destination points must match.
+        """
+
+        check_nD(src, 2)
+        check_nD(dst, 2)
+
+        if len(src) < 3 or len(dst) < 3:
+            raise ValueError(
+                "There should be at least 3 points in both sets (source and destination)."
+            )
+
+        if src.shape != dst.shape:
+            raise ValueError(
+                "The shape of source and destination coordinates must match."
+            )
+
+        self.src = src
+        n, d = src.shape
+
+        dist = sp.spatial.distance.cdist(self.src, self.src)
+        K = _radial_basis_kernel(dist)
+        P = np.hstack([np.ones((n, 1)), src])
+        L = np.zeros((n + 3, n + 3), dtype=np.float32)
+        L[:n, :n] = K
+        L[:n, -3:] = P
+        L[-3:, :n] = P.T
+        V = np.concatenate([dst, np.zeros((d + 1, d))])
+        self.spline_mappings = np.dot(np.linalg.inv(L), V)
```

Comment:
Explicitly inverting matrices is seldom a good idea. Use `linalg.solve`?

Reply:
Also, perhaps better not to reinvent the wheel and use SciPy's RBF to do the work?
Could reduce the code here by quite a bit, if it works.

Reply:
I remember looking at it in the past, but I don't remember why we didn't follow that approach. Totally possible, that we didn't yet have a good grasp on concepts. @decorouz, do you want to give it a shot?

Reply:
To add some context from a discussion @stefanv and me just had: `linalg.solve` is usually more precise in calculating the inverse, while `linalg.inv` is prone to small numeric errors that accumulate during the calculation. (At least that's what I understood :smiley:)

Reply:
And faster :)

The [short mathematical answer](https://math.stackexchange.com/a/1978293/656185).
The [longer mathematical unpacking](https://gregorygundersen.com/blog/2020/12/09/matrix-inversion/).